File: docker-environment/java-docker-demo/Logger.java
import com.google.gson.Gson;
import java.util.*;

public class Logger {
    private static List<Map<String, Object>> testCases = new ArrayList<>();
    private static int totalCorrect = 0;
    private static int totalTestcases = 0;

    public static void logTestCase(int index, TestRunner.TestCaseStats stats, List<Object> inputs, Object expectedOutput) {
        totalTestcases++;
        if (stats.passed) totalCorrect++;

        Map<String, Object> testCaseResult = new HashMap<>();
        testCaseResult.put("test_case", index);
        testCaseResult.put("passed", stats.passed);
        testCaseResult.put("inputs", new Gson().toJson(inputs));
        testCaseResult.put("output", stats.result == null ? "undefined" : new Gson().toJson(stats.result));
        testCaseResult.put("expected", new Gson().toJson(expectedOutput));
        testCaseResult.put("run_time", String.format("%.3f ms", stats.runTime));
        testCaseResult.put("memory", String.format("%.3f KB", stats.memoryUsed));

        testCases.add(testCaseResult);
    }

    public static void logSummary(TestRunner.Stats stats) {
        Map<String, Object> summary = new HashMap<>();
        summary.put("run_result", testCases);
        summary.put("total_correct", totalCorrect);
        summary.put("total_testcases", totalTestcases);
        summary.put("total_run_time", String.format("%.3f ms", stats.totalRunTime));
        summary.put("total_run_memory", String.format("%.3f MB", stats.totalMemoryUsed / 1024));
        summary.put("all_passed", stats.allTestsPassed);

        System.out.println(new Gson().toJson(summary));
    }
}

File: docker-environment/cpp-docker-demo/Logger.cpp
#include "Logger.h"
#include <iostream>
#include <iomanip>
#include <sstream>

double Logger::totalRuntime = 0.0;
double Logger::totalMemory = 0.0;
bool Logger::allTestsPassed = true;
int Logger::totalCorrect = 0;
int Logger::totalTestcases = 0;
json Logger::summaryJson;

void Logger::logTestCase(int index, const json& result, bool passed) {
    totalTestcases++;
    if (passed) totalCorrect++;

    json testCaseResult;
    testCaseResult["test_case"] = index;
    testCaseResult["passed"] = passed;
    testCaseResult["inputs"] = result["input"];
    testCaseResult["output"] = result["output"];
    testCaseResult["expected"] = result["expected"];

    std::stringstream ss;
    ss << std::fixed << std::setprecision(3) << result["runtime"].get<double>() / 1000.0;
    testCaseResult["run_time"] = ss.str() + " ms";

    ss.str("");  // 清空 stringstream
    ss << std::fixed << std::setprecision(3) << result["memory"].get<double>();
    testCaseResult["memory"] = ss.str() + " KB";

    summaryJson["run_result"].push_back(testCaseResult);

    totalRuntime += result["runtime"].get<double>() / 1000.0;
    totalMemory += result["memory"].get<double>();

    if (!passed) allTestsPassed = false;
}

void Logger::logSummary(const std::vector<json>& allResults) {
    summaryJson["total_correct"] = totalCorrect;
    summaryJson["total_testcases"] = totalTestcases;
    
    std::stringstream ss;
    ss << std::fixed << std::setprecision(3) << totalRuntime;
    summaryJson["total_run_time"] = ss.str() + " ms";
    
    ss.str("");  // 清空 stringstream
    ss << std::fixed << std::setprecision(3) << totalMemory / 1024.0;
    summaryJson["total_run_memory"] = ss.str() + " MB";
    
    summaryJson["all_passed"] = allTestsPassed;

    std::cout << summaryJson.dump() << std::endl;
}